src/viadot/sources/eurostat.py

```
@@ -0,0 +1,290 @@
+            requested_columns (List[str], optional): list of needed names of columns. Names should be given as str's into the list.
+                Defaults to None.
+            tests:
+                - `column_size`: dict{column: size}
+                - `column_unique_values`: list[columns]
+                - `column_list_to_match`: list[columns]
+                - `dataset_row_count`: dict: {'min': number, 'max', number}
+                - `column_match_regex`: dict: {column: 'regex'}
+                - `column_sum`: dict: {column: {'min': number, 'max': number}}
+        Raises:
+            APIError: If there is an error with the API request.
+            ValueError: If the resulting DataFrame is empty.
+            TypeError: If self.params is different type than a dictionary.
+            TypeError: If self.requested_columns is different type than a list containing strings.
+        Returns:
+            pd.DataFrame: Final DataFrame or raise prefect.logger.error, if issues occur.
+        """
+
+        self.dataset_code = dataset_code
+        self.params = params
+        if not isinstance(self.params, dict) and self.params is not None:
+            raise TypeError("Params should be a dictionary.")
+
+        self.base_url = f"{base_url}{self.dataset_code}?format=JSON&lang=EN"
+        self.requested_columns = requested_columns
+        if not isinstance(requested_columns, list) and requested_columns is not None:
+            raise TypeError("Requested columns should be provided as list of strings.")
+
+        # getting response from API
+        try:
+            response = handle_api_response(self.base_url, params=self.params)
+            data = response.json()
+            data_frame = self.eurostat_dictionary_to_df(["geo", "time"], data)
+        except APIError:
+            self.make_params_validation()
+
+        if data_frame.empty:
+            try:
+                self.make_params_validation()
```

Why not call this at the very beginning of `to_df()`? You don't need to call the API to know whether the params are correct or not (in fact thanks to this validation func, you know in advance if an API call will fail and so you can fail faster).

---

True, this could be better approach.

---

Hm actually I realized now you're actually calling the API every time to validate the params. But still, it seems like this belongs to `to_df()` as `params` should be a param of the `to_df()` method.